source/braille.py

```diff
@@ -1441,11 +1442,142 @@ def _setCursor(self, info: textInfos.TextInfo):
+			self._currentSelection = None
+			return api.getReviewPosition().copy()
+		info: textInfos.TextInfo
+		try:
+			info = self.obj.makeTextInfo(textInfos.POSITION_SELECTION)
+		except (LookupError, RuntimeError):
+			self._currentSelection = None
+			return self._collapsedReviewPosition()
+		if info.isCollapsed:
+			# Cursor
+			self._currentSelection = None
+			return self._collapsedReviewPosition()
+		# Selection
+		if self._currentSelection is None or self._currentSelection != info:
+			# Selection changed, update also review position
+			self._currentSelection = info.copy()
+			if self.obj.isTextSelectionAnchoredAtStart:
+				self._currentSelection.move(textInfos.UNIT_CHARACTER, -1, "end")
+			self._currentSelection.collapse(end=self.obj.isTextSelectionAnchoredAtStart)
+			# Enqueue to avoid recursion error
+			queueHandler.queueFunction(queueHandler.eventQueue, api.setReviewPosition, self._currentSelection)
+			self._currentSelection = info.copy()
+			return info
+		# Selection unchanged
+		readingUnit: textInfos.TextInfo = api.getReviewPosition().copy()
+		readingUnit.expand(self._getReadingUnit())
+		if (
+			readingUnit.start >= info.end
+			or readingUnit.end <= info.start
+		):
+			# Reading unit containing review position is outside of selection
+			return self._collapsedReviewPosition()
+		else:
+			# Reading unit contains selected characters
+			# All characters are not necessarily selected
+			if readingUnit.start < self._currentSelection.start:
+				readingUnit.start = self._currentSelection.start
+			if readingUnit.end > self._currentSelection.end:
+				readingUnit.end = self._currentSelection.end
+			self._readingUnitContainsSelectedCharacters = True
+			return readingUnit
+
+	def _collapsedReviewPosition(self) -> textInfos.TextInfo:
+		"""Gets collapsed review position.
+		:return: collapsed review position
+		"""
+		info: textInfos.TextInfo = api.getReviewPosition().copy()
+		# Info should be collapsed, but it is not always, at least when
+		# switching from focus mode to browse mode.
+		info.collapse()
+		return info
 
-	def _getSelection(self):
-		return api.getReviewPosition().copy()
+	def update(self):
+		"""Updates this region.
+		When "show selection" is enabled and reading unit contains selected
+		characters, adds functionality to properly move to the next and previous line,
+		to start and end of current control, to start and end of current line,
+		to scroll forward and back to the next and previous line,
+		and to move to focus.
+		"""
+		previousReadingUnit: textInfos.TextInfo | None = getattr(self, "_readingInfo", None)
+		super().update()
+		if (
+			not config.conf["braille"]["showSelection"]
+			or not self._readingUnitContainsSelectedCharacters
+			or self._currentScript is None
+			or previousReadingUnit is None
+		):
+			return
+		log.debug(f"jatketaan: {self._currentScript}")
+		from globalCommands import GlobalCommands
+		startOfNextOrPreviousLineScripts: set[_ScriptFunctionT] = {

```

Comment:
I don't believe this type hint is correct, `_ScriptFunctionT` matches a bound instance, i.e. `GlobalCommands().script_braille_nextLine`.

The items in this set really should be in the form `globalCommands.commands.script_braille_nextLine`.

Have you tested this? i.e. could you confirm the conditions like `self._currentScript in startOfNextOrPreviousLineScripts` evaluated correctly? because from my read these would never be True as `_currentScript` would never match an unbound `GlobalCommands` function.

Reply:
> I don't believe this type hint is correct, `_ScriptFunctionT` matches a bound instance, i.e. `GlobalCommands().script_braille_nextLine`.
> 
> The items in this set really should be in the form `globalCommands.commands.script_braille_nextLine`.
> 
> Have you tested this? i.e. could you confirm the conditions like `self._currentScript in startOfNextOrPreviousLineScripts` evaluated correctly? because from my read these would never be True as `_currentScript` would never match an unbound `GlobalCommands` function.

Yes I can confirm that conditions work. I think also that `type(self._currentScript)` is same as `type(GlobalCommands.script_braille_nextLine)`. `self._currentScript` gets its value from `scriptHandler.getCurrentScript()` which return type hint is `_ScriptFunctionT | None`.